### PR TITLE
Add -e/--environment/--environment-files to plan show/export

### PR DESCRIPTION
**Pull request by happz**
Only `tmt run` was allowed to accept environment options. That's fine until one wishes to run `tmt plan export` and expects various environment variables, used e.g. in `hardware` or `arch` keys, to be replaced.

Usualy, not really a problem, plan export simply emits what's been loaded from fmf, which is exactly what it should emit, but here comes Testing Farm: `plan export` is used for exposing `hardware` which then can be passed down to Artemis for provisioning.

Therefore the patch adds the same set of options to `plan show` and `plan export`, for those who would like to see how evaluated envvars affected the plan.

### CODE REVIEW

**Comment by happz** [comment]
Open for review, I don't plan any code changes, I will only add a test or two, maybe also a unit test for the new function. But the functionality is pretty much done AFAICT.

File: tmt/utils.py [hunk 1]
```diff
@@ -2026,6 +2026,113 @@ def environment_files_to_dict(
...
+    from_fmf_files: EnvironmentType = {}
+    from_fmf_dict: EnvironmentType = {}
+    from_cli_files: EnvironmentType = {}
+    from_cli: EnvironmentType = {}
+
+    if raw_fmf_environment_files is None:
+        pass
+
+    elif isinstance(raw_fmf_environment_files, list):
+        raw_fmf_environment_files = raw_fmf_environment_files or []
```
Comment (lukaszachy) [inline tmt/utils.py:2076]: Why is this necessary? Don't we already have a list instance?

File: tmt/utils.py [hunk 2]
```diff
@@ -2026,6 +2026,113 @@ def environment_files_to_dict(
...
+
+    else:
+        raise NormalizationError(
+            f'{key_address_prefix}environment', raw_fmf_environment, 'unset or a dictionary')
+
+    if raw_cli_environment_files is None:
+        pass
+
+    elif isinstance(raw_cli_environment_files, (list, tuple)):
+        raw_cli_environment_files = raw_cli_environment_files or []
```
Comment (lukaszachy) [inline tmt/utils.py:2105]: same Q: why is this line required ?

File: tmt/utils.py [hunk 3]
```diff
@@ -2026,6 +2026,113 @@ def environment_files_to_dict(
...
+
+    elif isinstance(raw_cli_environment, (list, tuple)):
+        from_cli = environment_to_dict(variables=list(raw_cli_environment), logger=logger)
+
+    else:
+        raise NormalizationError(
+            'environment', raw_cli_environment, 'unset or a list of key/value pairs')
+
+    # Combine all sources into one mapping. CLI gets precedence over fmf's `environment`,
+    # which takes precedence over
```
Comment (lukaszachy) [inline tmt/utils.py:2127]: unfinished sentence?

File: tmt/utils.py (hunk 1 above)
Comment (happz) [reply tmt/utils.py]: Correct, no longer needed, removed in b4776162

File: tmt/utils.py (hunk 3 above)
Comment (happz) [reply tmt/utils.py]: Fixed in b4776162

**Review by psss — APPROVED** [review]
Looks nice! Thanks for implementing this. I just can't resist one question in 821925a7 :-)

**Comment by happz** [comment]
Look at the recently added https://github.com/teemtee/tmt/blob/main/tmt/hardware.py, 1200 lines & 292 are empty :) Code like https://github.com/teemtee/tmt/blob/main/tmt/libraries/beakerlib.py#L189 is extremely hard for me to follow and understand - you will have a lot of fun removing those empty lines after me, I'm afraid I'm not going to give them up :)

**Comment by psss** [comment]
> I tend to separate logical steps with empty lines, which helps me follow the logical flow of the action and data - first I get my hands on the plan, then I take care of resolving the dynamic ref, then I would handle the exception, if there was one, etc.

Thanks much for the detailed explanation. It helped me understand your motivation for that (which I admit was a bit of mystery before). Grouping logical steps together makes a perfect sense to me. Perhaps it's just about the extent, granularity to which the separation is done. Which will be a matter of taste...

Speaking about the taste, in fact, the example you provided above (I mean the second one) is even 100% compatible with my taste as well ;-) But I'm afraid you would usually even separate that `if` and `else` clauses :)

> Look at the recently added https://github.com/teemtee/tmt/blob/main/tmt/hardware.py, 1200 lines & 292 are empty :)

Yes, yes, and it's a great source of inspiration... and wonder ;-)

> Code like https://github.com/teemtee/tmt/blob/main/tmt/libraries/beakerlib.py#L189 is extremely hard for me to follow and understand

Understood. Right, that method is not very beautiful. And first of all it's way too large. Perhaps comments sometimes serve a similar purpose as blank lines but some additional empty line can help a bit more.

I have to admit that you've already infected me with that white-space plague: When writing code I've noticed I'm adding much more empty lines than before. And cannot resist the feeling that the code somehow can breathe a bit more freely ;-)

**Comment by happz** [comment]
> Speaking about the taste, in fact, the example you provided above (I mean the second one) is even 100% compatible with my taste as well ;-) But I'm afraid you would usually even separate that `if` and `else` clauses :)

Yes, I do that :)

> I have to admit that you've already infected me with that white-space plague: When writing code I've noticed I'm adding much more empty lines than before. And cannot resist the feeling that the code somehow can breathe a bit more freely ;-)

Muhehe!!`</end-of-evil-laugh>`

Yeah, it's spacier, less dense. It would be much harder to read with an 80x25 terminal, but today... I guess one of the influences of the "functions shall be small and do one thing and one thing only" rule was also the visual limitation like this, it's hard to follow a long busy function when you can see 25 lines at once.

**Review by lukaszachy — APPROVED** [review]
